packages/protocol/contracts/L1/ProverPool.sol

```diff
@@ -0,0 +1,484 @@
+    // and change parameters.
+    // @Dani: Most prob. not for exit and modify, because it would require extra
+    // logic and cases. Let handle
+    // modification in a separate function
+    function stake(
+        uint32 totalAmount,
+        uint16 rewardPerGas,
+        uint16 capacity
+    )
+        external
+        nonReentrant
+    {
+        // TODO(daniel): assuming we have 1 block per second, and each proof
+        // takes
+        // 1 hour to generate, then we need a total capacity of 3600 at least
+        // across
+        // 32 provers, that means each prover's minimul capacity is 3600/32=112.
+        // We shall
+        require(capacity >= 100, "too small");
+
+        // Cannot enter the pool below the minimum
+        require(minimumStake < totalAmount, "Cannot enter below minimum");
+        require(rewardPerGas > 0, "Cannot be less than 1");
+        // Prover shall always be sender and in case of
+        // lending/cooperation/delegate
+        // it shall be done off-chain or at least outside of this pool contract.
+        address prover = _msgSender();
+        // The list of 32 is not full yet so kind of everybody can enter into
+        // the pool
+
+        // TODO: the logics should look like this:
+        // 1. Find the smallest staker in the top staker list, using its current
+        // stakedAmount
+        //    value, not the original stakedAmount value, say this is Alice.
+        // 2. Compare if the new staker, Bob's stakedAmount is bigger than
+        // Alice's. We may even compare
+        //    capacity (not sure how the math work), note that if Bob is already
+        // on the exiting list, we shall reuse its tokens locked there.
+        // 3. If Bob is perferred, we change Alice's status to `exiting` (as if
+        // Alice herself requested to exit)
+        // 4. We replace Alice with Bob in the `provers` list.
+        if (_anyEmptyStakingSlots()) {
+            proverData[currentIdIdx] = ProverData(
+                currentIdIdx, prover, 0, totalAmount, rewardPerGas, capacity
+            );
+            provers[currentIdIdx] = Prover(totalAmount, rewardPerGas);
+            proverToId[prover] = currentIdIdx;
+
+            // This would not overflow the 32 because of:
+            // provers[31].stakedAmount == 0
+            unchecked {
+                ++currentIdIdx;
+            }
+
+            // If we just filled up the last place in the array (so means full -
+            // we can
+            // determine which is the staker in risk)
+            if (provers[31].stakedAmount != 0) {
+                (minimumStakerId, minimumStake) = _determineLowestStaker();
+            }
+        } else {
+            // List is full, we need to check which position we need to put them
+            // into
+            (minimumStakerId,) = _determineLowestStaker();
+
+            // We need to overwrite it's place
+            proverData[minimumStakerId] = ProverData(
```

Comment:
Re-opening this question now since the file changed, which was:
- if i am the 32nd staker, and someone else comes in to replace me, you overwrite my place in this array. I am no longer in this array. I can not see what I have staked, nor my capacity/etc, nor can I even exit staking, since the `exit` functions look this up in the array.

My stake, since I have been replaced, is lost forever and I am unable to ever get it back. Even the record of what I staked is now gone. You have overwritten me.

There needs to be a way to keep track of ALL stakers, not just the top 32, or when replacing the 32nd person, you need to give them their stake back (which I like less).

As well, there is no way for me, as that now 33rd-infinite staker, to even see what my current stake IS. Maybe I only need to add some "X" TTKO to re-claim my spot in the top 32, but I am unable to fetch, from this contract, any info about my stake, as soon as I'm overwritten in this array.  

I hope this issue is explained better now than it was before. To put it simply, because you overwrite the entry in the array without respecting the original staker, that information and the ability to control the stake is lost forever, because you use `proverData[proverToId[_msgSender()]]` to control the stake in the future, but now, `proverData[myOriginalId]` is now a different staker.